src/main/java/net/minestom/server/inventory/click/ClickPreprocessor.java

```diff
@@ -0,0 +1,130 @@
+package net.minestom.server.inventory.click;
+
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.ints.IntLists;
+import net.minestom.server.entity.Player;
+import net.minestom.server.inventory.Inventory;
+import net.minestom.server.inventory.PlayerInventory;
+import net.minestom.server.network.packet.client.play.ClientClickWindowPacket;
+import net.minestom.server.utils.inventory.PlayerInventoryUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiFunction;
+
+/**
+ * Preprocesses click packets for an inventory, turning them into {@link ClickInfo} instances for further processing.
+ */
+public final class ClickPreprocessor {
+
+    private final @NotNull Inventory inventory;
```

Comment:
Does this class really need an `Inventory`? Would be great if it could become more standalone